addons/mass_mailing/controllers/main.py

```diff
@@ -22,6 +22,11 @@ def mailing(self, mailing_id, email=None, res_id=None, **post):
                 res_ids = contacts.ids
             else:
                 res_ids = [res_id]
+
+            # Do not save unsubscription if there is any failure or flaw
+            right_token = mailing._unsubscribe_token(res_id, email)
+            if token != right_token:
```

Comment:
Comparing security tokens should theoretically be done in a time-constant way, to avoid timing attacks. You should use `consteq` from the `odoo.tools` package to compare the tokens, like we're doing [elsewhere](https://github.com/odoo/odoo/blob/46be8b30d39de9abaeaa282458d17008d52ccf1d/addons/mail/controllers/main.py#L37).
While we're dealing with security, we might as well do it right from the start ;-)